### ISSUE TITLE
Build failure with GHC 9.4

### ISSUE THREAD

**Issue by andreasabel** [martijnbastiaan/doctest-parallel#42]
It would be nice to have a version of `doctest-parallel` soon so that the ecosystem can more quickly embrace GHC 9.4.

Building with GHC 9.4 and Cabal-3.8 fails for version 0.2.3:
```
Building library for doctest-parallel-0.2.3..
[ 1 of 14] Compiling Language.Haskell.GhciWrapper ( ghci-wrapper/src/Language/Haskell/GhciWrapper.hs, dist/build/Language/Haskell/GhciWrapper.o, dist/build/Language/Haskell/GhciWrapper.dyn_o )
[ 2 of 14] Compiling Paths_doctest_parallel ( dist/build/autogen/Paths_doctest_parallel.hs, dist/build/Paths_doctest_parallel.o, dist/build/Paths_doctest_parallel.dyn_o )
[ 3 of 14] Compiling Test.DocTest.Helpers ( src/Test/DocTest/Helpers.hs, dist/build/Test/DocTest/Helpers.o, dist/build/Test/DocTest/Helpers.dyn_o )

src/Test/DocTest/Helpers.hs:35:48: error:
    Module
    ‘Distribution.PackageDescription.Parsec’
    does not export
    ‘readGenericPackageDescription’
   |
35 | import Distribution.PackageDescription.Parsec (readGenericPackageDescription)
   |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ 4 of 14] Compiling Test.DocTest.Internal.GhcUtil ( src/Test/DocTest/Internal/GhcUtil.hs, dist/build/Test/DocTest/Internal/GhcUtil.o, dist/build/Test/DocTest/Internal/GhcUtil.dyn_o )
[ 5 of 14] Compiling Test.DocTest.Internal.Interpreter ( src/Test/DocTest/Internal/Interpreter.hs, dist/build/Test/DocTest/Internal/Interpreter.o, dist/build/Test/DocTest/Internal/Interpreter.dyn_o )
[ 6 of 14] Compiling Test.DocTest.Internal.Location ( src/Test/DocTest/Internal/Location.hs, dist/build/Test/DocTest/Internal/Location.o, dist/build/Test/DocTest/Internal/Location.dyn_o )
[ 7 of 14] Compiling Test.DocTest.Internal.Options ( src/Test/DocTest/Internal/Options.hs, dist/build/Test/DocTest/Internal/Options.o, dist/build/Test/DocTest/Internal/Options.dyn_o )
[ 8 of 14] Compiling Test.DocTest.Internal.Util ( src/Test/DocTest/Internal/Util.hs, dist/build/Test/DocTest/Internal/Util.o, dist/build/Test/DocTest/Internal/Util.dyn_o )
[ 9 of 14] Compiling Test.DocTest.Internal.Extract ( src/Test/DocTest/Internal/Extract.hs, dist/build/Test/DocTest/Internal/Extract.o, dist/build/Test/DocTest/Internal/Extract.dyn_o )

src/Test/DocTest/Internal/Extract.hs:187:27: error:
    • Couldn't match expected type: Maybe a0 -> IO HscEnv
                  with actual type: IO HscEnv
    • The function ‘initializePlugins’
      is applied to two value arguments,
        but its type ‘HscEnv -> IO HscEnv’ has only one
      In the first argument of ‘liftIO’, namely
        ‘(initializePlugins hsc_env Nothing)’
      In a stmt of a 'do' block:
        hsc_env' <- liftIO (initializePlugins hsc_env Nothing)
    |
187 |       hsc_env' <- liftIO (initializePlugins hsc_env Nothing)
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Test/DocTest/Internal/Extract.hs:247:31: error:
    Variable not in scope: unpackHDS :: HsDoc GhcPs -> String
    Suggested fix:
      Perhaps use one of these:
        ‘unpackHDSC’ (imported from GHC),
        ‘unpackFS’ (imported from GHC.Data.FastString)
    |
247 |   map (fmap (toLocated . fmap unpackHDS)) docs
    |                               ^^^^^^^^^

src/Test/DocTest/Internal/Extract.hs:252:24: error:
    • Couldn't match type: GenLocated SrcSpan (HsDoc GhcPs)
                     with: WithHsDocIdentifiers HsDocString GhcPs
      Expected: [(Maybe String, LHsDoc GhcPs)]
        Actual: [(Maybe String, GenLocated SrcSpan (LHsDoc GhcPs))]
    • In the first argument of ‘(++)’, namely ‘exports’
      In the second argument of ‘(++)’, namely ‘exports ++ decls’
      In the expression: header ++ exports ++ decls
    |
252 |   docs     = header ++ exports ++ decls
    |                        ^^^^^^^

src/Test/DocTest/Internal/Extract.hs:252:35: error:
    • Couldn't match type ‘HsDocString’
                     with ‘WithHsDocIdentifiers HsDocString GhcPs’
      Expected: [(Maybe String, LHsDoc GhcPs)]
        Actual: [(Maybe String, LHsDocString)]
    • In the second argument of ‘(++)’, namely ‘decls’
      In the second argument of ‘(++)’, namely ‘exports ++ decls’
      In the expression: header ++ exports ++ decls
    |
252 |   docs     = header ++ exports ++ decls
    |                                   ^^^^^

src/Test/DocTest/Internal/Extract.hs:306:3: error:
    • The constructor ‘HsPar’ should have 4 arguments, but has been given 2
    • In the pattern: HsPar _ (L l e)
      In a \case alternative: HsPar _ (L l e) -> extractLit (locA l) e
      In the expression:
        \case
          HsPar _ (L l e) -> extractLit (locA l) e
          ExprWithTySig _ (L l e) _ -> extractLit (locA l) e
          HsOverLit _ OverLit {ol_val = HsIsString _ s}
            -> Just (toLocated (L loc (unpackFS s)))
          HsLit _ (HsString _ s) -> Just (toLocated (L loc (unpackFS s)))
          _ -> Nothing
    |
306 |   HsPar _ (L l e) -> extractLit (locA l) e
    |   ^^^^^^^^^^^^^^^

src/Test/DocTest/Internal/Extract.hs:351:31: error:
    • Expecting one more argument to ‘DocDecl’
      Expected a type, but ‘DocDecl’ has kind ‘* -> *’
    • In the type signature:
        fromDocDecl :: SrcSpan -> DocDecl -> (Maybe String, LHsDocString)
      In an equation for ‘extractDocStrings’:
          extractDocStrings
            = everythingBut
                (++)
                (([], False) `mkQ` fromLHsDecl `extQ` fromLDocDecl
                   `extQ` fromLHsDocString)
            where
                fromLHsDecl :: DocSelector (LHsDecl GhcPs)
                fromLHsDecl (L loc decl)
                  = case decl of
                      DocD _ x -> select (fromDocDecl (locA loc) x)
                      _ -> ...
                fromLDocDecl :: DocSelector (LDocDecl GhcPs)
                fromLDocDecl (L loc x) = select (fromDocDecl (locA loc) x)
                ....
    |
351 |     fromDocDecl :: SrcSpan -> DocDecl -> (Maybe String, LHsDocString)
    |                               ^^^^^^^
Error: cabal: Failed to build doctest-parallel-0.2.3
```

**Comment by andreasabel** [comment]
Added the bound Cabal < 3.8 to 0.2.3 on hackage:  https://hackage.haskell.org/package/doctest-parallel-0.2.3/revisions/

**Comment by martijnbastiaan** [comment]
Sure, I'll have a look.

> Added the bound Cabal < 3.8 to 0.2.3 on hackage: https://hackage.haskell.org/package/doctest-parallel-0.2.3/revisions/

Could you edit the file in this repo too? That would make sure the changes get propagated to the new release too.

**Comment by andreasabel** [comment]
> > Added the bound Cabal < 3.8 to 0.2.3 on hackage: [hackage.haskell.org/package/doctest-parallel-0.2.3/revisions](https://hackage.haskell.org/package/doctest-parallel-0.2.3/revisions/)
> 
> Could you edit the file in this repo too? That would make sure the changes get propagated to the new release too.

I would hope that the new release works with `Cabal-3.8` which is shipped with GHC 9.4, so this bound would not be necessary then (nor helpful).

**Comment by martijnbastiaan** [comment]
Well it's up to the maintainers of packages to decide whether they'd like to release their next version with GHC 9.4 support, so I feel like a PR or at least a separate issue would be quite courteous :).

In any case, I've added GHC 9.4 support. It should land in `main` soon!

**Comment by martijnbastiaan** [comment]
GHC 9.4 has been published in `doctest-parallel-0.2.4`.
